### PR TITLE
Support "near" parameter in prepared query service block

### PR DESCRIPTION
**Pull request by ryanuber**
Allows the use of the `Near` parameter in a prepared query definition. This parameter works just like the `?near=foo` HTTP API parameter, only it is baked into the query, allowing it to be used from the DNS interface. 

This enables some interesting functionality and opens up additional useful prepared query features when using DNS as the query mechanism.

/cc @slackpad

### CODE REVIEW

File: consul/prepared_query_endpoint.go [hunk 1]
```diff
@@ -368,10 +368,44 @@ func (p *PreparedQuery) Execute(args *structs.PreparedQueryExecuteRequest,
...
+	}
+
+	// Respect the magic "_agent" flag.
+	preferLocal := false
+	if sortFrom.Node == "_agent" {
+		preferLocal = true
+		sortFrom = args.Origin
+	}
+
+	// Perform the distance sort
```
Comment (sean-) [inline consul/prepared_query_endpoint.go]: Do we want to memoize this value for ~60s or whatever the interval is for the tomography calcs?  I seem to recall someone saying this calc was "expensive," but if it's not, disregard.

File: consul/structs/prepared_query.go [hunk 2]
```diff
@@ -173,6 +178,12 @@ type PreparedQueryExecuteRequest struct {
 	// Limit will trim the resulting list down to the given limit.
 	Limit int
 
+	// Origin is used to carry around a reference to the node which
+	// is executing the query on behalf of the client. It is taken
+	// as a QuerySource so that it can be used directly for queries
+	// relating to the agent servicing the request.
+	Origin QuerySource
```
Comment (slackpad) [inline consul/structs/prepared_query.go]: Did you take a look at the existing `Source` parameter right after this one? I think it was originally intended to do what `Origin` does.

File: command/agent/http.go [hunk 3]
```diff
@@ -531,13 +531,7 @@ func (s *HTTPServer) parseToken(req *http.Request, token *string) {
 // DC in the request, if given, or else the agent's DC.
 func (s *HTTPServer) parseSource(req *http.Request, source *structs.QuerySource) {
 	s.parseDC(req, &source.Datacenter)
-	if node := req.URL.Query().Get("near"); node != "" {
```
Comment (slackpad) [inline command/agent/http.go]: I think this'll break other endpoints that depend on `?near=_agent` getting resolved at this level. I think you can pass the existing `Source` member of the execute request into here.

File: consul/prepared_query_endpoint.go [hunk 4]
```diff
@@ -368,10 +368,44 @@ func (p *PreparedQuery) Execute(args *structs.PreparedQueryExecuteRequest,
...
+	if err := p.srv.sortNodesByDistanceFrom(sortFrom, reply.Nodes); err != nil {
 		return err
 	}
 
+	// Nodes cannot be any "closer" than localhost, so this special case ensures
+	// the local node is returned first if it is present in the result. This
+	// allows the local agent to be preferred even when network coordinates are
+	// not enabled. Only works if the results come from the request origin DC.
+	if preferLocal && reply.Datacenter == args.Origin.Datacenter {
```
Comment (slackpad) [inline consul/prepared_query_endpoint.go]: I don't think the special case is worth it since 0.6 has been out for a long time. This way you won't need to know about `_agent` at all at this level.

File: consul/prepared_query_endpoint.go (hunk 1 above)
Comment (slackpad) [inline consul/prepared_query_endpoint.go]: It's not too bad - I'd probably leave as-is since this can be DNS TTL-ed and read-scaled with staleness across the servers.

File: consul/prepared_query_endpoint.go [hunk 5]
```diff
@@ -368,10 +368,44 @@ func (p *PreparedQuery) Execute(args *structs.PreparedQueryExecuteRequest,
...
+		preferLocal = true
+		sortFrom = args.Origin
+	}
+
+	// Perform the distance sort
+	if err := p.srv.sortNodesByDistanceFrom(sortFrom, reply.Nodes); err != nil {
 		return err
 	}
 
+	// Nodes cannot be any "closer" than localhost, so this special case ensures
```
Comment (slackpad) [inline consul/prepared_query_endpoint.go]: This doesn't seem worth the special case since 0.6 has been out for a while and nobody should be disabling coordinates :-) It'll do some stuff in the common case where the thing is already in the front position as well, and that would add extra complexity to avoid.

**Comment by slackpad** [comment]
@ryanuber took a look through and added some comments - this is looking good! I think the `.Source` thing is the biggest change since that's already there, otherwise some minor stuff.

File: consul/structs/prepared_query.go (hunk 2 above)
Comment (ryanuber) [inline consul/structs/prepared_query.go]: Pretty close yeah. I originally tried to use that, but I switched to this because in the HTTP/DNS api code at the agent, you can't be sure what the prepared query actually does, so the agent can't be sure if it should pass its own name + datacenter, or the user's provided `?near` and/or `?dc` value. This just makes it so we can send both and decide outside of the agent.

File: command/agent/http.go (hunk 3 above)
Comment (ryanuber) [inline command/agent/http.go]: Does it break? This just pushes the check out to the `consul` side instead of the `agent`. I was trying to avoid checking the magic flag in the agent and on the consul side (since the `Near` parameter can live in either place now).

Comment (slackpad) [inline command/agent/http.go]: Still think you'll need to revert this change - should not be needed any more and there are other RPC endpoints that do the sort which rely on `_agent` getting resolved on the agent side, not the server side.

File: consul/prepared_query_endpoint.go [hunk 6]
```diff
@@ -368,7 +368,25 @@ func (p *PreparedQuery) Execute(args *structs.PreparedQueryExecuteRequest,
 	// Shuffle the results in case coordinates are not available if they
 	// requested an RTT sort.
 	reply.Nodes.Shuffle()
-	if err := p.srv.sortNodesByDistanceFrom(args.Source, reply.Nodes); err != nil {
+
+	// Build the query source. This can be provided by the client, or by
+	// the prepared query. Client-specified takes priority.
+	qs := args.Source
+	if qs.Datacenter == "" {
```
Comment (slackpad) [inline consul/prepared_query_endpoint.go]: DC should always be supplied by `parseDC` but this is fine and defensive.

File: consul/structs/prepared_query.go [hunk 7]
```diff
@@ -34,6 +34,11 @@ type ServiceQuery struct {
 	// discarded)
 	OnlyPassing bool
 
+	// Near allows the query to always prefer the node nearest the given
+	// node. If the node does not exist, results are returned in their
+	// normal randomly-shuffled order.
```
Comment (slackpad) [inline consul/structs/prepared_query.go]: Would be good to mention the magic `_agent` value here in the comment.

**Comment by slackpad** [comment]
@ryanuber looks good! I still think we need to revert the `parseSource()` changes and the only other thing is to mention this in the docs. Otherwise, :shipit:.

File: command/agent/http.go (hunk 3 above)
Comment (ryanuber) [inline command/agent/http.go]: Oh I see, catalog and health both allow distance sorting. Will fix!

**Comment by slackpad** [comment]
:+1:
